plugins/org.eclipse.n4js/src/org/eclipse/n4js/utils/FindReferenceHelper.java

```diff
@@ -0,0 +1,123 @@
+import org.eclipse.xtext.resource.IResourceDescriptions;
+import org.eclipse.xtext.resource.impl.ResourceDescriptionsProvider;
+
+import com.google.common.collect.Lists;
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+
+/**
+ * Helper class to find references.
+ */
+@SuppressWarnings("restriction")
+public class FindReferenceHelper {
+
+	@Inject
+	private Provider<TargetURIs> targetURISetProvider;
+
+	@Inject
+	private IReferenceFinder referenceFinder;
+
+	@Inject
+	private TargetURICollector collector;
+
+	@Inject
+	private ResourceDescriptionsProvider resourceDescriptionsProvider;
+
+	/** @return all references to the given declaration. Respect editor states. */
+	public List<EObject> findReferences(EObject declaration) {
+		declaration = getDeclaration(declaration);
+		TargetURIs targets = getTargets(declaration);
+		Resource eResource = declaration.eResource();
+		SimpleResourceAccess resourceAccess = new SimpleResourceAccess(eResource.getResourceSet());
+		IResourceDescriptions index = resourceDescriptionsProvider.getResourceDescriptions(eResource);
+		ReferenceAcceptor acceptor = new ReferenceAcceptor();
+
+		referenceFinder.findAllReferences(targets, resourceAccess, index, acceptor, null);
+
+		return acceptor.results;
+	}
+
+	/** @return all references to the given declaration in the given {@link Resource} */
+	public List<EObject> findReferencesInResource(EObject declaration, Resource resource) {
+		declaration = getDeclaration(declaration);
+		TargetURIs targets = getTargets(declaration);
+		ReferenceAcceptor acceptor = new ReferenceAcceptor();
+
+		referenceFinder.findReferences(targets, resource, acceptor, null);
+
+		return acceptor.results;
+	}
+
+	private EObject getDeclaration(EObject declaration) {
+		if (declaration instanceof ParameterizedTypeRef) {
+			declaration = ((ParameterizedTypeRef) declaration).getDeclaredType();
+		}
+		return declaration;
+	}
+
+	private TargetURIs getTargets(EObject declaration) {
+		// Special handling for composed members
+		List<EObject> realTargets = new ArrayList<>();
+		if ((declaration instanceof TMember) && ((TMember) declaration).isComposed()) {
+			// In case of composed member, add the constituent members instead.
+			List<TMember> constituentMembers = ((TMember) declaration).getConstituentMembers();
+			for (TMember constituentMember : constituentMembers) {
+				realTargets.add(constituentMember);
+			}
+		} else {
+			// Standard case
+			realTargets.add(declaration);
+		}
+
+		TargetURIs targets = targetURISetProvider.get();
+
+		for (EObject realTarget : realTargets) {
+			collector.add(realTarget, targets);
+		}
+		return targets;
+	}
+
+	static class ReferenceAcceptor implements IReferenceFinder.Acceptor {
+		final ArrayList<EObject> results = Lists.newArrayList();
+
+		@Override
+		public void accept(EObject src, URI srcURI, EReference eRef, int idx, EObject tgtOrProxy, URI tgtURI) {
+			results.add(src);
+		}
+
+		@Override
```

Comment:
We could probably remove this second `accept`  method

Reply:
Apparently, we are implementing an Interface here, so we need to declare this method.